### PR TITLE
Filter Logical Replication Stream by INSERT/UPDATE/DELETE/TRUNCATE

### PR DESCRIPTION
**Pull request by barakshechter**
Adding an --operations command line option to select which operation types to be replicated to the target kinesis stream.

### CODE REVIEW

**Comment by barakshechter** [comment]
We have a need to use logical replication to stream changes from our database to other parts of our application, but only need information on deleted records. This PR would add the ability to only replicate certain operation types from the logical replication stream to the target kinesis stream.

Any feedback you might have is appreciated. Thanks!

**Comment by barakshechter** [comment]
@surbas Not sure if you saw this PR opened. Can we do anything to help push this forward?

**Comment by surbas** [comment]
Hey sorry @barakshechter . Been heads down. Promise to give this a look by end of week.

File: pg2kinesis/formatter.py [hunk 1]
```diff
@@ -104,7 +105,7 @@ def _preprocess_wal2json_change(self, change):
 
         change_dictionary = json.loads(change)
         if not change_dictionary:
-            return None
+            return []
```
Comment (surbas) [inline pg2kinesis/formatter.py]: Why this change?

File: pg2kinesis/formatter.py [hunk 2]
```diff
@@ -143,7 +144,7 @@ def __call__(self, change):
         return [self.produce_formatted_message(pp_change) for pp_change in pp_changes]
 
     def produce_formatted_message(self, change):
-        return change
+        raise NotImplementedError
```
Comment (surbas) [inline pg2kinesis/formatter.py]: Why this change?

File: pg2kinesis/formatter.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 # Tuples representing changes as pulled from database
 Change = namedtuple('Change', 'xid, table, operation, pkey')
 FullChange = namedtuple('FullChange', 'xid, change')
+FullChange.operation = property(lambda self: self.change.get('kind'))
```
Comment (surbas) [inline pg2kinesis/formatter.py]: Where is this used?

File: pg2kinesis/__main__.py [hunk 4]
```diff
@@ -84,6 +110,9 @@ def __call__(self, change):
         progress_msg = 'xid: {:12} win_count:{:>10} win_size:{:>10}mb cum_count:{:>10} cum_size:{:>10}mb'
 
         for fmt_msg in fmt_msgs:
+            if not self.should_send_to_kinesis(fmt_msg):
+                fmt_msg = None
+
             did_put = self.writer.put_message(fmt_msg)
```
Comment (surbas) [inline pg2kinesis/__main__.py]: Do we put_message with 'None' as the message?

File: pg2kinesis/__main__.py [hunk 5]
```diff
@@ -32,9 +51,12 @@
               help='Attempt to on start create a the slot.')
 @click.option('--recreate-slot', default=False, is_flag=True,
               help='Deletes the slot on start if it exists and then creates.')
+@click.option('--operations', default='a', type=parse_operations_param,
```
Comment (surbas) [inline pg2kinesis/__main__.py]: I think we can use built in `choice` type in `click` with `multiple` set to `true`, instead of needing this custom type.

See: https://stackoverflow.com/questions/37486367/click-choice-for-multiple-arguments

File: pg2kinesis/formatter.py (hunk 3 above)
Comment (barakshechter) [reply pg2kinesis/formatter.py]: This is used when determining what operation this change corresponds to, so we can decide whether it should be filtered out. The base `Change` object exposes this as `change.operation`, so this creates a shorthand for `full_change.operation`. 

This allows us to avoid something like:
```
if instanceof(change, Change):
    change.operation
elif instanceof(change, FullChange):
    change.change.get('kind')
else:
    ???
```
and instead just do `if change.operation in operations_to_replicate:...`.

File: pg2kinesis/__main__.py (hunk 4 above)
Comment (barakshechter) [reply pg2kinesis/__main__.py]: yes, i believe this was already being done to ensure that the replication slot is flushed. 

with filtering out operations, this is really important, because you could be filtering out most operations, which if we didn't ensure flushing on a regular basis, would lead to the replication slot growing unchecked. It must be periodically flushed to avoid eating up the entire free disk space.

File: pg2kinesis/__main__.py (hunk 5 above)
Comment (barakshechter) [reply pg2kinesis/__main__.py]: I used the custom parser so we could do something more streamlined like `--operations ui` instead of `--operations update insert`, but i can change

File: pg2kinesis/formatter.py (hunk 1 above)
Comment (barakshechter) [reply pg2kinesis/formatter.py]: this is used like this:
```
        pp_changes = self._preprocess_wal2json_change(change)
        return [self.produce_formatted_message(pp_change) for pp_change in pp_changes]
```

if you return None, this will end up raising an exception, whereas if there are no changes to process, returning an empty list will be a no-op and wait for the next change in the repslot.

**Comment by surbas** [comment]
THanks for your thoughtful answers, and the new feature!
